horace_core/sqw/@sqw/parse_pixel_indices.m

```diff
@@ -1,212 +1,206 @@
-if nargin==1
-    if nout>0
-        irun = cell(size(win));
-        idet = cell(size(win));
-        ien  = cell(size(win));
-        for i = 1:nw
-            if iscell(win)
-                pix = win{i}.pix;
-                runid_map = win{i}.runid_map;
-            else
-                pix = win(i).pix;
-                runid_map = win(i).runid_map;
-            end
-
-            run_idx= pix.run_idx';
-            irun{i} = arrayfun(@(x)runid_map(x),run_idx);
-            idet{i} = pix.detector_idx';
-            ien{i}  = pix.energy_idx';
+sz_win = size(win);
+
+% If ipix is given, check that it is consistent with the number and sizes of the
+% input sqw object(s)
+ipix_present = exist('ipix', 'var');
+if ipix_present
+    % Get min and max indices in ipix (as column vectors)
+    % If ipix is numeric, then get make it a cell array that is replicated to the
+    % size of the input sqw object array. This is to make the block of code that
+    % gets the output for a single sqw object much cleaner (without lots of
+    % capturing of cases of numeric or cell array input, and scalar or array
+    % cell arrays). The creation of a cell that is then input to repmat has no
+    % memory or CPU penalty as the contents are not altered at any point, so it
+    % is essentially just making a bunch of pointers.
+    if iscell(ipix)
+        % Assume cell array of integer arrays - none of them are permitted to be empty
+        n_ipix = numel(ipix);
+        if n_ipix==0 || any(cellfun(@isempty, ipix))
+            error('HORACE:parse_pixel_indices:invalid_argument', ['The cell array ',...
+                '''ipix'' is empty, or one or more arrays in ''ipix'' is empty'])
+        end
+        if ~any(numel(ipix) == [1,nw])
+            error('HORACE:parse_pixel_indices:invalid_argument', ['The number of ',...
+                'arrays in ''ipix'' must be 1 or the number of sqw objects: %d'], nw)
+        end
+        if numel(ipix)==1
+            min_ipix = min(ipix{1}(:)) * ones(sz_win);
+            max_ipix = max(ipix{1}(:)) * ones(sz_win);
+            ipix_cellarray = repmat(ipix, sz_win);          % effectively a set of pointers
+        else
+            min_ipix = cellfun(@(x)(min(x(:))), ipix(:));   % output is column vector
+            max_ipix = cellfun(@(x)(max(x(:))), ipix(:));
+            ipix_cellarray = reshape(ipix, sz_win);
+        end
+    else
+        % Assume ipix is an integer array - but it cannot be empty
+        if isempty(ipix)
+            error('HORACE:parse_pixel_indices:invalid_argument', ...
+                'Optional pixel index argument ''ipix'' is empty')
         end
-
         if nw == 1
-            irun = irun{1};
-            idet = idet{1};
-            ien = ien{1};
+            min_ipix = min(ipix(:));
+            max_ipix = max(ipix(:));
+            ipix_cellarray = {ipix};
+        else
+            min_ipix = min(ipix(:)) * ones(size(win));
+            max_ipix = max(ipix(:)) * ones(size(win));
```

Comment:
```suggestion
            min_ipix = repmat(min(ipix(:)), size(win));
            max_ipix = repmat(max(ipix(:)), size(win));
```

Reply:
I've always assumed that repmat is a more expensive operation than ones. I went online and there is a surprising amount of debate about using a * ones(size(M)), a + zeros(size(M)), repmat (a, size(M)). WIthout spending more than a couple of minutes on it, it would appear there is much of a muchness regarding timings, although my guess would be repmat has more overheads because it has to check the type of what is being repeated. Personally, I find a * ones(size(M) easiest to understand, so am leaving it as is. Would be interesting for us to have a discussion one day...